confgenerator/resourcedetector/gce_metadata_provider.go

```diff
@@ -0,0 +1,92 @@
+package resourcedetector
+
+import (
+	"fmt"
+	"strings"
+
+	gcp_metadata "cloud.google.com/go/compute/metadata"
+)
+
+const notAvailable = "NOT AVAILABLE"
+
+// Provide the GCE metadata using the on-VM metadata server
+type GCEMetadataProvider struct {
+	client *gcp_metadata.Client
+}
+
+func NewGCEMetaDataProvider() GCEDataProvider {
+	c := gcp_metadata.NewClient(nil)
+	return &GCEMetadataProvider{c}
+}
+
+func (gmp *GCEMetadataProvider) getProject() (string, error) {
+	return gmp.client.ProjectID()
+}
+
+func (gmp *GCEMetadataProvider) getZone() (string, error) {
+	return gmp.client.Zone()
+}
+
+// We assume the current running instance has at least one network interface
+// Otherwise won't be able to connect to
+func (gmp *GCEMetadataProvider) getNetwork() (string, error) {
+	return gmp.client.Get("instance/network-interfaces/0/network")
+}
+
+func (gmp *GCEMetadataProvider) getSubnetwork() (string, error) {
+	return notAvailable, nil
+}
+
+func (gmp *GCEMetadataProvider) getPublicIP() (string, error) {
+	return gmp.client.ExternalIP()
+}
+
+func (gmp *GCEMetadataProvider) getPrivateIP() (string, error) {
+	return gmp.client.InternalIP()
+}
+
+func (gmp *GCEMetadataProvider) getInstanceID() (string, error) {
+	return gmp.client.InstanceID()
+}
+
+func (gmp *GCEMetadataProvider) getInstanceName() (string, error) {
+	return gmp.client.InstanceName()
+}
+
+func (gmp *GCEMetadataProvider) getTags() (string, error) {
+	tags, err := gmp.client.InstanceTags()
+	if err != nil {
+		return "", err
+	}
+	return strings.Join(tags, ","), nil
+}
+
+func (gmp *GCEMetadataProvider) getMetadataKeys() (string, error) {
+	return gmp.client.Get("instance/attributes")
+}
+
+func (gmp *GCEMetadataProvider) getMetadata(key string) (string, string, error) {
+	v, err := gmp.client.Get(fmt.Sprintf("instance/attributes/%s", key))
+	return key, v, err
+}
+
+// GCE VM labels are currently not available in the metadata server
+func (gmp *GCEMetadataProvider) getLabel() (string, error) {
+	return notAvailable, nil
+}
+
+func (gmp *GCEMetadataProvider) getMachineType() (string, error) {
+	return gmp.client.Get("instance/machine-type")
+}
+
+func (gmp *GCEMetadataProvider) getInterfaceNames() (string, error) {
+	return gmp.client.Get("instance/network-interfaces/")
+}
+
+// The metadata server would return interfaces as "0/", needs to trim the "/"
+// Also add `nic` to the name following the Prometheus's SD convention
+func (gmp *GCEMetadataProvider) getInterfaceIPv4(name string) (string, string, error) {
+	interface_name := strings.TrimSuffix(name, "/")
+	ipv4, err := gmp.client.Get(fmt.Sprintf("instance/network-interfaces/%s/ip", interface_name))
```

Comment:
These are very interesting mutations we have to do. I'd leave it out of the detector package and add it when the `prometheus` receiver uses it. Only it cares about the Prometheus SD's conventions.